modules/dxp/apps/multi-factor-authentication/multi-factor-authentication-fido2-web/src/main/resources/META-INF/resources/js/AuthenticationTransformer.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * The contents of this file are subject to the terms of the Liferay Enterprise
+ * Subscription License ("License"). You may not use this file except in
+ * compliance with the License. You can obtain a copy of the License by
+ * contacting Liferay, Inc. See the License for the specific language governing
+ * permissions and limitations under the License, including but not limited to
+ * distribution rights of the Software.
+ */
+
+import * as webauthn from './webauthn';
+
+export default function AuthenticationTransformer({
+	additionalProps,
+	portletNamespace,
+	...otherProps
+}) {
+	return {
+		...otherProps,
+		onClick() {
+			let {assertionRequest} = additionalProps;
+
+			if (!assertionRequest) {
+				return;
+			}
+
+			assertionRequest = JSON.parse(assertionRequest);
+
+			webauthn
+				.get(assertionRequest.publicKeyCredentialRequestOptions)
+				.then((value) => {
+					const publicKey = webauthn.credentialToObject(value);
+
+					if (
+						publicKey.response.userHandle !== null &&
+						publicKey.response.userHandle === ''
+					) {
+						delete publicKey.response.userHandle;
+					}
+
+					const responseInput = document.getElementById(
+						`${portletNamespace}responseJSON`
+					);
+
+					responseInput.value = JSON.stringify(publicKey);
+				})
+				.catch(() => {
+					const messageContainer = document.getElementById(
+						`${portletNamespace}messageContainer`
+					);
+
+					messageContainer.innerHTML = `<span class="alert alert-danger">
+                    ${Liferay.Language.get(
+						'your-authenticator-was-unable-to-verify-your-credential'
+					)}
+                </span>`;
+				});
```

Comment:
This also just feels odd to modify an element rather than fire off some sort of error event or call back with the content of the error.

Then again, I may just be in the react world for too long where things like this don't feel normal. So take my comments with a grain of salt.

Reply:
> Then again, I may just be in the react world for too long where things like this don't feel normal.

:joy_cat: 

The way I see it, (and I might be wrong) is that for the moment we are in this "in between" dimension where React component and non React live together, sometimes they hang out and do these kind of odd things.

Jokes aside, if there's a way to improve this (be it now or later), I'm all for it.

Reply:
I think you're totally right about living in between, and I think these sort of `*Transformer` files are more likely to live in this in between world more than most other files.

I just need to take off my react glasses from time to time 😂